icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java

```diff
@@ -0,0 +1,740 @@
+                    this.getSingleUnits()) {
+                result.add(new MeasureUnitImpl(singleUnit));
+            }
+
+            return result;
+        }
+
+        result.add(this.clone());
+        return result;
+    }
+
+    /**
+     * Applies dimensionality to all the internal single units.
+     * For example: <b>square-meter-per-second</b>, when we apply dimensionality -2, it will be <b>square-second-per-p4-meter</b>
+     */
+    public void applyDimensionality(int dimensionality) {
+        for (SingleUnitImpl singleUnit :
+                singleUnits) {
+            singleUnit.setDimensionality(singleUnit.getDimensionality() * dimensionality);
+        }
+    }
+
+    /**
+     * Mutates this MeasureUnitImpl to append a single unit.
+     *
+     * @return true if a new item was added. If unit is the dimensionless unit,
+     * it is never added: the return value will always be false.
+     */
+    public boolean appendSingleUnit(SingleUnitImpl singleUnit) {
+        identifier = null;
+
+        if (singleUnit.isDimensionless()) {
+            // We don't append dimensionless units.
+            return false;
+        }
+
+        // Find a similar unit that already exists, to attempt to coalesce
+        SingleUnitImpl oldUnit = null;
+        for (int i = 0, n = this.singleUnits.size(); i < n; i++) {
+            SingleUnitImpl candidate = this.singleUnits.get(i);
+            if (candidate.isCompatibleWith(singleUnit)) {
+                oldUnit = candidate;
+                break;
+            }
+        }
+
+        if (oldUnit != null) {
+            // Both dimensionalities will be positive, or both will be negative, by
+            // virtue of isCompatibleWith().
+            oldUnit.setDimensionality(oldUnit.getDimensionality() + singleUnit.getDimensionality());
+
+            return false;
+        }
+
+        // TODO: shall we just add singleUnit instead of creating a copy ??
+        this.singleUnits.add(singleUnit.clone());
+
+        // If the MeasureUnitImpl is `UMEASURE_UNIT_SINGLE` and after the appending a unit, the singleUnits are more
+        // than one singleUnit. thus means the complexity should be `UMEASURE_UNIT_COMPOUND`
+        if (this.singleUnits.size() > 1 && this.complexity == MeasureUnit.Complexity.SINGLE) {
+            this.setComplexity(MeasureUnit.Complexity.COMPOUND);
+        }
+
+        return true;
+    }
+
+    /**
+     * Transform this MeasureUnitImpl into a MeasureUnit, simplifying if possible.
+     */
+    public MeasureUnit build() {
+        return new MeasureUnit(this);
+    }
+
+    /**
+     * @return SingleUnitImpl
+     * @throws UnsupportedOperationException if the object could not be converted to SingleUnitImpl.
+     */
+    public SingleUnitImpl getSingleUnitImpl() {
+        if (this.singleUnits.size() == 0) {
+            return new SingleUnitImpl();
+        }
+        if (this.singleUnits.size() == 1) {
+            return this.singleUnits.get(0).clone();
```

Comment:
Looking at the call sites for this function, you don't always need to clone.  For example, in getDimensionality, you are only reading fields, not writing fields.  Consider moving the `.clone()` out into the call sites that need it.

Reply:
I think cloning single unit is pretty cheap and this will prevent any conflict in the future, what do you think ?

Reply:
I think you should move it out, but not strongly.  I'll resolve.